host/windows/syscall.c

```diff
@@ -312,9 +450,94 @@ int oe_syscall_rmdir_ocall(const char* pathname)
 **==============================================================================
 */
 
+#define OE_SOCKET_FD_MAGIC 0x29b4a345c7564b57
+typedef struct win_socket_fd
+{
+    uint64_t magic;
+    SOCKET socket;
+} oe_socket_fd_t;
+
+static oe_socket_fd_t _invalid_socket = {OE_SOCKET_FD_MAGIC, INVALID_SOCKET};
+
+oe_host_fd_t _make_socket_fd(SOCKET sock)
+{
+    oe_host_fd_t fd = (oe_host_fd_t)&_invalid_socket;
+    if (sock != INVALID_SOCKET)
+    {
+        oe_socket_fd_t* socket_fd =
+            (oe_socket_fd_t*)malloc(sizeof(oe_socket_fd_t));
+        socket_fd->magic = OE_SOCKET_FD_MAGIC;
+        socket_fd->socket = sock;
+        fd = (oe_host_fd_t)socket_fd;
+    }
+    return fd;
+}
+
+SOCKET _get_socket(oe_host_fd_t fd)
+{
+    oe_socket_fd_t* socket_fd = (oe_socket_fd_t*)fd;
+    if (socket_fd && socket_fd->magic == OE_SOCKET_FD_MAGIC)
+        return socket_fd->socket;
+    return INVALID_SOCKET;
+}
+
+static oe_host_fd_t _dup_socket(oe_host_fd_t oldfd)
+{
+    oe_socket_fd_t* old_socket_fd = (oe_socket_fd_t*)oldfd;
+    if (old_socket_fd && old_socket_fd->magic == OE_SOCKET_FD_MAGIC)
+    {
+        oe_socket_fd_t* new_socket_fd =
+            (oe_socket_fd_t*)malloc(sizeof(oe_socket_fd_t));
+        if (new_socket_fd)
+        {
+            *new_socket_fd = *old_socket_fd;
+            return (oe_host_fd_t)(uint64_t)new_socket_fd;
+        }
+        else
+        {
+            _set_errno(OE_ENOMEM);
+        }
+    }
+    return -1;
+}
+
+static int _wsa_startup()
+{
+    static bool wsa_init_done = FALSE;
+    WSADATA wsaData;
+    int ret = 0;
+
+    if (!wsa_init_done)
```

Comment:
this code looks ok for TrustZone which is single-threaded, but does not look thread-safe for SGX

Reply:
Used oe_atomic_compare_and_swap